contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -2,18 +2,141 @@
 
 #[allow(unused_imports)]
 use gstd::prelude::*;
+use gstd::{exec, msg, debug};
+use tamagotchi_nft_io::{Tamagotchi, TmgAction};
+
+static mut TAMAGOTCHI_STATE: Option<Tamagotchi> = None;
+
+const HUNGER_PER_BLOCK: u64 = 1;
+const BOREDOM_PER_BLOCK: u64 = 2;
+const ENERGY_PER_BLOCK: u64 = 2;
+const FILL_PER_FEED: u64 = 1000;
+const FILL_PER_ENTERTAINMENT: u64 = 1000;
+const FILL_PER_SLEEP: u64 = 1000;
 
 #[no_mangle]
 extern fn init() {
-    // TODO: 0️⃣ Copy the `init` function from the previous lesson and push changes to the master branch
+    let name_bytes = msg::load_bytes().expect("Invalid initial name");
+    let name = String::from_utf8(name_bytes).expect("Invalid UTF-8");
+
+    let date_of_birth = exec::block_timestamp();
+
+    let tamagotchi = Tamagotchi {
+        name,
+        date_of_birth,
+        owner: msg::source(),
+        fed: 1000,
+        fed_block: 0,
+        entertained: 1000,
+        entertained_block: 0,
+        slept: 1000,
+        slept_block: 0,
+        approved_account: None,
+    };
+
+    save_tamagotchi_state(tamagotchi);
+
+    debug!("Tamagotchi initialized with name: {:?}, date of birth: {:?}", name, date_of_birth);
 }
 
 #[no_mangle]
 extern fn handle() {
-    // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    let action: TmgAction = msg::load().expect("Failed to decode TmgAction");
+
+    let current_block_height = exec::block_height() as u64;
+    let tamagotchi = unsafe {
+        TAMAGOTCHI_STATE
+            .as_mut()
+            .expect("The contract is not initialized")
+    };
+
+    update_levels(tamagotchi, current_block_height);
+
+    match action {
+        TmgAction::Name => {
+            msg::reply(&tamagotchi.name, 0).expect("Failed to send reply");
+        },
+        TmgAction::Age => {
+            let current_timestamp = exec::block_timestamp();
+            let age_in_milliseconds = current_timestamp - tamagotchi.date_of_birth;
+            msg::reply(&age_in_milliseconds, 0).expect("Failed to send reply");
+        },
+        TmgAction::Feed => {
+            tamagotchi.fed_block = current_block_height;
+            tamagotchi.fed = tamagotchi.fed.saturating_add(FILL_PER_FEED);
+        },
+        TmgAction::Entertain => {
+            tamagotchi.entertained_block = current_block_height;
+            tamagotchi.entertained = tamagotchi.entertained.saturating_add(FILL_PER_ENTERTAINMENT);
+        },
+        TmgAction::Sleep => {
+            tamagotchi.slept_block = current_block_height;
+            tamagotchi.slept = tamagotchi.slept.saturating_add(FILL_PER_SLEEP);
+        },
+        TmgAction::Transfer(new_owner) => {
+            tamagotchi.owner = new_owner;
+            msg::reply("Transferred", 0).expect("Failed to send reply");
+        },
+        TmgAction::Approve(account) => {
+            tamagotchi.approved_account = Some(account);
```

Comment:
The same: You must check that the msg::source() is an owner! Otherwise anyone can make an approval

Reply:
The validation has been added